File: compiler/ast/statements.c

#include "statements.h"

#define NEW_NODE(var_name, name) struct name_node* var_name = malloc(sizeof(struct name_node))
#define RETURN_NODE(var_name) (struct node*) var_name;
#define SET_BASE(var_name, t) node->base = (struct node*){.type = t}

CREATE_NODE(for_stat)(struct statement_node* init, struct expression_node* cond,
    struct expression_node* post, struct block_stat_node block){

    NEW_NODE(node, for_stat);

    node->init = init;
    node->cond = cond;
    node->post = post;
    node->block = block;
    SET_BASE(node, NODE_FOR_STAT);

    RETURN_NODE(node);
}

CREATE_NODE(return_stat)(struct expression_node** results){

    NEW_NODE(node, return_stat);

    node->results = results;
    SET_BASE(node, NODE_RETURN_STAT);

    RETURN_NODE(node);
}

File: compiler/ast/statements.go
package ast

import (
	"github.com/end-r/guardian/compiler/lexer"
)

type AssignmentStatementNode struct {
	Modifiers []lexer.TokenType
	Left      []ExpressionNode
	Operator  lexer.TokenType
	Right     []ExpressionNode
}

func (n AssignmentStatementNode) Type() NodeType { return AssignmentStatement }

type ReturnStatementNode struct {
	Results []ExpressionNode
}

func (n ReturnStatementNode) Type() NodeType { return ReturnStatement }

type ConditionNode struct {
	Condition ExpressionNode
	Body      *ScopeNode
}

func (n ConditionNode) Type() NodeType { return IfStatement }

type IfStatementNode struct {
	Init       Node
	Conditions []ConditionNode
	Else       *ScopeNode
}

func (n IfStatementNode) Type() NodeType { return IfStatement }

type SwitchStatementNode struct {
	Target      ExpressionNode
	Cases       *ScopeNode
	IsExclusive bool
}

func (n SwitchStatementNode) Type() NodeType { return SwitchStatement }

type CaseStatementNode struct {
	Expressions []ExpressionNode
	Block       *ScopeNode
}

func (n CaseStatementNode) Type() NodeType { return CaseStatement }

type ForStatementNode struct {
	Init  *AssignmentStatementNode
	Cond  ExpressionNode
	Post  StatementNode
	Block *ScopeNode
}

func (n ForStatementNode) Type() NodeType { return ForStatement }
